rasa/core/evaluation/marker_tracker_loader.py

```diff
@@ -0,0 +1,72 @@
+import random
+from rasa.shared.core.trackers import DialogueStateTracker
+from typing import Any, Iterable, List, Text, Optional
+from rasa.core.tracker_store import TrackerStore
+
+
+class MarkerTrackerLoader:
+    """Represents a wrapper around a `TrackerStore` that loads trackers in a configurable access pattern."""
+
+    @staticmethod
+    def strategy_all(count: int, keys: Iterable[Text]) -> Iterable[Text]:
+        """Selects all keys from the set of keys."""
+        return keys
+
+    @staticmethod
+    def strategy_first_n(count: int, keys: Iterable[Text]) -> Iterable[Text]:
+        """Takes the first N keys from the set of keys."""
+        return keys[:count]
+
+    @staticmethod
+    def strategy_sample(count: int, keys: Iterable[Text]) -> Iterable[Text]:
+        """Takes a sample of N keys from the set of keys."""
+        return random.choices(keys, k=count)
+
+    _STRATEGY_MAP = {
+        "all": strategy_all,
+        "first_n": strategy_first_n,
+        "sample": strategy_sample,
+    }
+
+    def __init__(
+        self,
+        tracker_store: TrackerStore,
+        strategy: str,
+        count: int = None,
+        seed: Any = None,
+    ) -> None:
+        """Create a MarkerTrackerLoader.
+        
+        Args:
+            tracker_store: The underlying tracker store to access.
+            strategy: The strategy to use for selecting trackers, can be 'all', 'sample', or 'first_n'.
+            count: Number of trackers to return, can only be None if strategy is 'all'.
+            seed: Optional seed to set up random number generator, only useful if strategy is 'sample'.
+        """
+        self.tracker_store = tracker_store
+        self.count = count
+
+        if strategy not in MarkerTrackerLoader._STRATEGY_MAP:
+            raise  # Not sure what best exception type to use here is, subclass of RasaException?
+
+        self.strategy = MarkerTrackerLoader._STRATEGY_MAP[strategy]
+
+        if not count:
+            if strategy != "all":
+                raise  # Exception
+            else:
+                pass  # Raise warning - redundant parameter
+
+        if seed:
+            if strategy == "sample":
+                random.seed(seed)
+            else:
+                pass  # Warning - redundant parameter
+
+    # Not super happy with this - is it sufficient to pull out to optionals or do we unwrap
+    # If we unwrap, do we need to make up the numbers
+    # Do we want a generator version that yields one at a time to reduce memory demands?
+    def load(self) -> List[Optional[DialogueStateTracker]]:
```

Comment:
So there's a question here about if we want to move this to using a generator - so then we fetch one tracker at a time which constrains memory usage (and if done right maybe gives us some better options for parallelism). Also want to know if we want these unwrapped from the `Optional` when returning, and if so do we want to make up the numbers if some aren't found (they should in theory always be found but we should handle this case if we're unwrapping) or just let the returned number vary from the desired number (what we manage to unwrap is what we get)

Reply:
I should say I think this should be a fairly trivial change, it's just which way do we prefer? After sleeping on it I think I prefer the generator approach since it's a bit more flexible

Reply:
>and if so do we want to make up the numbers if some aren't found

Can you clarify what you mean by the numbers? 

I think it's a good idea to parallelise loading the trackers and extracting the markers. We can start with a non-parallelised version for simplicity and add this if we have time, unless it's trivial to do. What do you think?

Reply:
> Can you clarify what you mean by the numbers?
By make up the numbers I mean if we sample 100 Optionals, unwrap them all and we have 10 Nones, we only have 90 samples - do we want to just log that we've lost 10 (implies some weird storage issue, probably easiest) or we could attempt to oversample and then just take 100 (more faff, not guaranteed to get 100 either but probably more chance of it), or do we want to fail out and exit?

So making it a generator should be trivial, parallelising it from the generator I think is straightforward too but would happen elsewhere/be a later addition. I was thinking I could change it now and then the type signature should remain the same (barring any weird things appearing), which would make those modifications down the line hopefully a little easier